src/map/roe.h

```diff
@@ -0,0 +1,118 @@
+/*
+ * roe.h
+ *      Author: Kreidos | github.com/kreidos
+ *
+===========================================================================
+
+  Copyright (c) 2020 Topaz Dev Teams
+
+  This program is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see http://www.gnu.org/licenses/
+
+===========================================================================
+*/
+
+#ifndef SRC_MAP_ROE_H_
+#define SRC_MAP_ROE_H_
+
+#include <array>
+#include <bitset>
+#include <vector>
+
+#include "../common/cbasetypes.h"
+#include "ai/helpers/event_handler.h"
+#include "packets/weather.h"
+
+struct lua_State;
+
+class CItemContainer;
+
+class CBaseEntity;
+
+enum ROE_EVENT
+{
+    ROE_MOBKILL = 1,
+    ROE_WSKILL_USE = 2,
+    ROE_LOOTITEM = 3,
+    ROE_SYNTHSUCCESS = 4,
+    ROE_DMGTAKEN = 5,
+    ROE_DMGDEALT = 6,
+    ROE_NONE // End of enum marker and OOB checkpost. Do not move or remove, place any new types above.
+};
+
+struct RoeCache
+{
+        std::bitset<4096> ImplementedRecords;
+        std::bitset<4096> RepeatableRecords;
+};
+
+struct RoeCheckHandler
+{
+    std::bitset<4096> bitmap;
+};
+
+extern std::array<RoeCheckHandler, ROE_NONE> RoeHandlers;
+
+enum class RoeDatagramPayload
+{
+    mob,
+    uinteger,
+};
+
+struct RoeDatagram
+{
+    RoeDatagramPayload type;
+    std::string param;
+    union data {
```

Comment:
I'm fine with this implementation, so no need to change it, but this would be a good place to use `std::variant`:
- All the types are the same size
- You're using a `type` key to pull out what you want and do different things in a switch in `roeutils::event(...)`
- Mabe everything is looking like a nail because I'm currently holding the `std::variant` hammer

Reply:
Ha ha, noted. In a future phase, type-safing `union` to `variant` may be prudent I just know and liked the struct/union paradigm; you reach for the tools you know best. :shrug: (Fee Fi Fo Fum, I smell C programmers!)